<ConceptMap /> SVG, Highlighting, CSS, oh my!

neenjaw opened the pull request:
Fixes https://github.com/exercism/v3-website/issues/202

This is a bigger PR comprising a new feature for the concept map and a significant refactor to incorporate transition fading.

1) New feature: Highlight all ancestors

When hovering over a concept, it now highlights all ancestors required to unlock that concept and the paths taken to get there.

2) With this new feature in place it became apparent that the previous strategy of a single canvas to render all of the paths was becoming too difficult to extend and maintain.  **Why?** There was a split-second flash when paths overlapped concepts and caused a rendering gaffe. The gaffe is because the fading in of the paths was instantaneous and the concept boxes faded. To manage independent fade animations would be reinventing the wheel.  So rather than having a single canvas, now a multi-canvas approach is taken so that the fading can all be managed with CSS.  From my research (googling SO) this shouldn't be a performance issue.

The only downside is that it doesn't have fine-grain control over the layering, so at this point the paths may overlay each other unpredictably.

---

There is also a minor refactor to align ConceptState to ConceptStatus in the react component

neenjaw committed:
controller updated to new interface
now uses UserTrack::GenerateConceptStatusMapping to create
status mapping

neenjaw committed:
ancestors now highlight on concept hover

neenjaw committed:
Tidying up active and inactive concept states

neenjaw committed:
Single canvas approach changed to multi-canvas

neenjaw committed:
ConceptState -> ConceptStatus

neenjaw committed:
remove debug console messages

neenjaw commented on app/javascript/components/concept-map/ConceptConnections.tsx:
```diff
@@ -1,92 +1,96 @@
+      canvas.style.height = `${height}px`
+
+      const paths = mapToPaths(connections)
+
```
```suggestion
```

ErikSchierboom commented on app/javascript/components/concept-map/Concept.tsx:
```diff
@@ -11,26 +11,21 @@ export const Concept = ({
   isActive: boolean
-  isDimmed: boolean
-  adjacentConcepts: string[]
+  isInactive: boolean
```
Can a concept be both active _and_ inactive?

neenjaw replied:
No, I'm using these flags to determine highlighting and local opacity, I can change the naming so it is more clear because there are 3 states. Neutral (when nothing highlighted), active (hovering) and inactive (dimmed)

ErikSchierboom replied:
That would be great. Maybe introduce an enum for that?

SleeplessByte replied:
`'neutral' | 'active' | 'dimmed'`

Fine to use the booleans for now. (Still wanted to say this, even if it's outdated)

ErikSchierboom reviewed (commented):
I was trying to test this locally, but I keep getting the following error (on all tracks I tested: Elixir, F#, Ruby):

```
Uncaught TypeError: Cannot read property 'start' of undefined
    at scalePath (draw-helpers.ts:175)
    at drawPath (draw-helpers.ts:25)
    at ConceptConnections.tsx:60
    at Array.forEach (<anonymous>)
    at ConceptConnections.tsx:28
    at commitHookEffectListMount (react-dom.development.js:19662)
    at commitPassiveHookEffects (react-dom.development.js:19699)
    at HTMLUnknownElement.callCallback (react-dom.development.js:191)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:240)
    at invokeGuardedCallback (react-dom.development.js:293)
```

iHiD reviewed (changes requested):
I get the same issue as Erik.

ErikSchierboom reviewed (approved):
It now works on my machine!

neenjaw committed:
Large refactor for performance increase
Several changes have been made to the component for how it handles the
path generation and management.

Before the canvas was sized to the webpage, it is no longer. It is now
sized to just the size needed to hold the path shape.  then a CSS
transform is applied to translate the canvas to the correct location.

Second performance was the elimination of extra redrawing.  The crux of
the problem is that the concepts don't know about the paths and the
paths don't know about the concepts, but depend on the concepts being
mounted to the DOM.  So a "channel" for communicating updates is now
used to allow concepts to self-register their location and for paths to
declare their need for them without causing extra renders.

So now, on load there is one DOM paint for the elements, and one canvas
paint per canvas to show the paths.

Some concerns exist whether this approach will work in the longer run,
so an SVG approach should probably be explored.

Infinite gratitude to @iHiD and @SleeplessByte for their help

Co-authored-by: Jeremy Walker <jez.walker@gmail.com>
Co-authored-by:
Derk-Jan Karrenbeld <derk-jan+github@karrenbeld.info>

neenjaw committed:
fix nullish coalescing for undefined

neenjaw committed:
fix webpage size handler from causign extra render

neenjaw committed:
add visibility handler

ErikSchierboom reviewed (approved):
Great! It now renders really fast on my machine.

neenjaw committed:
MASSIVE REFACTOR
Now uses SVG like it should have all along. 🤯

types cleaned up, files cleaned up, still more to do for that I'm sure

neenjaw committed:
fix test import, let to const

neenjaw committed:
svg styling

neenjaw committed:
add tailwind negative zindexes

neenjaw commented on app/controllers/tracks/concepts_controller.rb:
```diff
@@ -9,7 +9,9 @@ def authenticated_index
-      @concept_map_data[:status] = UserTrack::GenerateConceptStatusMapping.(@user_track)
+      @concept_map_data = Track::DetermineConceptMapLayout.(@track).merge(
+        status: UserTrack::GenerateConceptStatusMapping.(@user_track)
+      )
```
```suggestion
      @concept_map_data[:status] = UserTrack::GenerateConceptStatusMapping.(@user_track)
```
I think there was a merge conflict and I did something weird here, this is to fix.

iHiD replied:
It should be this:
```suggestion
      @concept_map_data[:status] = UserTrack::GenerateConceptStatusMapping.(@user_track)
```

iHiD replied:
(I've merged your suggestion before, but then it got overriden, so I'm leaving it so you can fix it 🙂)

neenjaw replied:
I feel like I've made this same change several times, I don't know where the commit keeps getting dropped :grinning:

neenjaw committed:
add module boundary function return types

neenjaw committed:
Add render when custom font loads.
As it is, the rendering concept map is faster than the loading the
custom font "Poppins". When the font loads, the font is swapped causing
the concept's DOM width to change but this doesn't trigger a react
render.  This adds the useFontLoaded hook so it will trigger a reload
when the font load event is detected.

iHiD commented on app/javascript/components/concept-map/Concept.tsx:
```diff
@@ -10,30 +17,37 @@ export const Concept = ({
   classes.push(`card-${status}`)
-  if (isActive) {
-    classes.push('card-active')
+  if (isInactive) {
```
I think this should still be `isActive`. Having negatively-named variables is generally a bit of a smell. Even though you're only checking for negation, it's better to have `if(!active)` than `if(inactive)`.

neenjaw replied:
Fixed in update

iHiD commented on app/javascript/components/concept-map/helpers/style-helpers.ts:
```diff
@@ -0,0 +1,40 @@
+    | 'line-width'
+    | 'circle-radius'
+    | 'hover-opacity'
+) => computedRootStyle.getPropertyValue(`--c-concept-map-${type}`)
```
So probably all of these can just be removed now we're just using the stylesheet.

neenjaw replied:
Yes, those have been removed.

neenjaw committed:
Optimize all of the things!
- Memoization added to several aspects so that it only recalculates what
changes.
- Request animation frame wrappers added to redrawing calls.

- some type cleanup
- unused function cleanup

iHiD commented on app/javascript/components/concept-map/PathLineSVG.tsx:
```diff
@@ -1,10 +1,21 @@
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [path.start.x, path.start.y, path.end.x, path.end.y])
+
+  return <path d={computedBezier} className={`line-width-2 ${path.status}`} />
```
I'd rather the `line-width-2` can be added via the stylesheet, rather than inlined.

I'm also not sure the `path.status` needs adding here. We can instead use the status reference that appears on the SVG itself for styling changes.

iHiD replied:
```suggestion
  return <path d={computedBezier} />
```

iHiD commented on app/javascript/components/concept-map/ConnectionPathSVG.tsx:
```diff
@@ -0,0 +1,109 @@
+          cx={normalizedPath.start.x}
+          cy={normalizedPath.start.y}
+          radius={radius}
+          status={normalizedPath.status}
```
I don't think this needs the status passing in. We can infer styling from the top-level SVG.

iHiD replied:
```suggestion
```

neenjaw replied:
True, Just need to go the next step of refactoring the stylesheet, thanks for the direction!

iHiD commented on app/javascript/components/concept-map/ConnectionPathSVG.tsx:
```diff
@@ -0,0 +1,109 @@
+          cx={normalizedPath.end.x}
+          cy={normalizedPath.end.y}
+          radius={radius}
+          status={normalizedPath.status}
```
I don't think this needs the status passing in. We can infer styling from the top-level SVG.

iHiD replied:
```suggestion
```

iHiD commented on app/javascript/components/concept-map/ConnectionPathSVG.tsx:
```diff
@@ -0,0 +1,109 @@
+    !existsActivePaths ||
+    (activeConcepts.has(connection.from) && activeConcepts.has(connection.to))
+
+  const classNames = ['canvas', normalizedPath.status]
```
We should probably rename this from `canvas` to `connection` to be semantically descriptive.

iHiD replied:
```suggestion
  const classNames = ['canvas', path.status]
```

iHiD commented on app/javascript/components/concept-map/ConceptMap.css:
```diff
@@ -59,7 +68,54 @@
+        & .end-cap {
+            fill: var(--c-concept-map-card-background);
+            stroke-dasharray: var(--c-concept-map-line-dasharray-solid);
+        }
```
```suggestion
          stroke-width: 2px;
        }
```

neenjaw replied:
I didn't need it here, it's being inherited from elsewhere

iHiD commented on app/javascript/components/concept-map/ConceptMap.css:
```diff
@@ -59,7 +68,54 @@
+            stroke-dasharray: var(--c-concept-map-line-dasharray-solid);
+        }
+
+        & .line-width-2 {
```
```suggestion
        & path {
```

iHiD commented on app/javascript/components/concept-map/PathLineEndSVG.tsx:
```diff
@@ -0,0 +1,25 @@
+      cx={cx}
+      cy={cy}
+      r={radius}
+      className={`line-width-2 end-cap ${status}`}
```
```suggestion
```

iHiD commented on app/javascript/components/concept-map/ConceptMap.css:
```diff
@@ -59,7 +68,54 @@
+            stroke-dasharray: var(--c-concept-map-line-dasharray-solid);
+        }
+
+        & .end-cap {
```
```suggestion
        & circle {
```

neenjaw replied:
Rather than styling the element, I've used another named class to prevent potential cascade rule problems

iHiD commented on app/javascript/components/concept-map/helpers/path-helpers.ts:
```diff
@@ -1,82 +1,46 @@
+      x: leftToRight ? width - radius - lineWidth : radius + lineWidth,
+      y: height - radius - lineWidth,
+    },
+    status: path.status,
```
```suggestion
```

neenjaw replied:
This change requires the type to change which cascades to a bunch of other changes.

Perhaps the better solution is to refactor so that the generated original path **is** the normalized path.

neenjaw committed:
fix controller

neenjaw committed:
Changes based on review from @iHiD
type ConceptPath is now ConceptPathProperties, which is computede by
`computePathProperties`. Led to a refactor since we don't need to maintain the
absolute coordinates for the svg path any more, and can just use the
size of the computed viewbox to determine the start and end based
on the coordinates of the concept boxes.

Also several CSS improvements.
